### ISSUE TITLE
Conflicting suggestions for unit_arg lint between clippy and (nightly) rustc

### ISSUE THREAD

**Issue by vitalyd** [rust-lang/rust-clippy#2945]
Given the following (contrived) code:
```rust
#![allow(unused)]
use std::io;

fn unit_fn() -> io::Result<()> {
    Ok(())
}

fn fallible() -> io::Result<()> {
    Ok(unit_fn()?)
}

fn main() {}
```
[Playground](https://play.rust-lang.org/?gist=e2dbb849ddbf8d83857fbd55edc00b8e&version=nightly&mode=debug&edition=2015)

clippy lints this as:
```
   Checking playground v0.0.1 (file:///playground)
warning: passing a unit value to a function
 --> src/main.rs:9:8
  |
9 |     Ok(unit_fn()?)
  |        ^^^^^^^^^^
  |
  = note: #[warn(unit_arg)] on by default
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#unit_arg
help: if you intended to pass a unit value, use a unit literal instead
  |
9 |     Ok(())
  |        ^^

    Finished dev [unoptimized + debuginfo] target(s) in 1.35s
```
If I intentionally change the code to the following (incorrect) version:
```rust
#![allow(unused)]
use std::io;

fn unit_fn() -> io::Result<()> {
    Ok(())
}

fn fallible() -> io::Result<()> {
    unit_fn()?
}

fn main() {}
```
I get a compiler error (on current nightly):
```
   Compiling playground v0.0.1 (file:///playground)
error[E0308]: try expression alternatives have incompatible types
 --> src/main.rs:9:5
  |
9 |     unit_fn()?
  |     ^^^^^^^^^^
  |     |
  |     expected enum `std::result::Result`, found ()
  |     help: try wrapping with a success variant: `Ok(unit_fn()?)`
  |
  = note: expected type `std::result::Result<(), std::io::Error>`
             found type `()`
```
So it seems like rustc is suggesting a form that clippy lints against, i.e.:
```
help: try wrapping with a success variant: `Ok(unit_fn()?)`
```
Note that the `stable` compiler does not have that `help: try wrapping ...` suggestion, so it's something recently added.

Also, to be clear - I know the code is incorrect and can be written differently.  My only point is to illustrate the discrepancy between syntactic/stylistic forms that clippy and rustc suggest.

**Comment by Manishearth** [comment]
yeah we just shouldn't be warning if `?` is involved

**Comment by vitalyd** [comment]
@Manishearth, I agree - I actually came up with this contrived example because I saw this lint fire in some real code, and started playing around with it because I didn't think it should fire at all as I like the `Ok(fallible_unit_fn()?)` formulation 😄
